File: PYTHON/MaxBuildingHeight1840.py
'''
Problem Statement:
You want to build n new buildings in a city labeled from 1 to n.
- The height of each building must be a non-negative integer.
- The height of the first building must be 0.
- The height difference between any two adjacent buildings cannot exceed 1.
- You are given a 2D integer array restrictions where restrictions[i] = [idi, maxHeighti].
Return the maximum possible height of the tallest building.
'''

'''
Approach: Two-Pass Optimization (Similar to Candy / Trapping Rain Water)
1. Boundary Extensions: Add a pseudo-restriction at building 1 with height 0, and at building n 
   with a theoretical max height of n - 1 (since it grows at most 1 per step from building 1).
2. Sorting: Sort the restrictions based on building IDs so we can propagate limits linearly.
3. Forward Pass: Update each building's max height constraint by checking how high it can go 
   starting from its left neighbor: r[i][1] = min(r[i][1], r[i-1][1] + distance).
4. Backward Pass: Pass backwards to propagate constraints from right neighbors back to left: 
   r[i][1] = min(r[i][1], r[i+1][1] + distance).
5. Peak Maximization: Between any two restricted buildings, the height can grow to a peak 
   and drop down. The theoretical peak height between two points (id1, h1) and (id2, h2) is 
   given by the math formula: floor((h1 + h2 + (id2 - id1)) / 2). Track the maximum peak found.
'''
class MaxBuildingHeight1840:
    def maxBuilding(self, n: int, restrictions: list[list[int]]) -> int:
        # Append boundary conditions to complete limits definition
        # Building 1 must have height 0; Building n can at most scale up to n - 1
        restrictions.append([1, 0])
        restrictions.append([n, n - 1])
        
        # Sort values based on the building index positions
        restrictions.sort(key=lambda x: x[0])
        
        m = len(restrictions)
        
        # Pass 1: Left-to-Right push to cap limits based on leftward bottlenecks
        for i in range(1, m):
            dist = restrictions[i][0] - restrictions[i - 1][0]
            restrictions[i][1] = min(restrictions[i][1], restrictions[i - 1][1] + dist)
            
        # Pass 2: Right-to-Left pull to tighten restrictions based on rightward bottlenecks
        for i in range(m - 2, -1, -1):
            dist = restrictions[i + 1][0] - restrictions[i][0]
            restrictions[i][1] = min(restrictions[i][1], restrictions[i + 1][1] + dist)
            
        max_height = 0
        
        # Final Sweep: Deduce optimal peaks generated between adjacent restriction points
        for i in range(m - 1):
            id1, h1 = restrictions[i]
            id2, h2 = restrictions[i + 1]
            
            # Intersection peak math formula: peak = ((h1 + h2) + distance) // 2
            peak = (h1 + h2 + (id2 - id1)) // 2
            max_height = max(max_height, peak)
            
        return max_height

if __name__ == '__main__':
    try:
        n_val = int(input("Enter total number of buildings n: "))
        m_val = int(input("Enter number of restrictions: "))
        
        restrictions_arr = []
        print("Enter restrictions (id height) line by line:")
        for _ in range(m_val):
            row = list(map(int, input().split()))
            restrictions_arr.append(row)
            
        solver = MaxBuildingHeight1840()
        result = solver.maxBuilding(n_val, restrictions_arr)
        print("Maximum possible building height:", result)
    except ValueError:
        print("Invalid input format.")

File: JAVA/Leetcode/MaxBuildingHeight1840.java
/*
 * Problem Statement:
 * You want to build n new buildings in a city labeled from 1 to n.
 * - The height of each building must be a non-negative integer.
 * - The height of the first building must be 0.
 * - The height difference between any two adjacent buildings cannot exceed 1.
 * - You are given a 2D integer array restrictions where restrictions[i] = [idi, maxHeighti].
 * Return the maximum possible height of the tallest building.
 */

/*
 * Approach: Two-Pass Optimization (Similar to Candy / Trapping Rain Water)
 * 1. Boundary Extensions: Add a pseudo-restriction at building 1 with height 0, and at building n 
 * with a theoretical max height of n - 1 (since it grows at most 1 per step from building 1).
 * 2. Sorting: Sort the restrictions based on building IDs so we can propagate limits linearly.
 * 3. Forward Pass: Update each building's max height constraint by checking how high it can go 
 * starting from its left neighbor: r[i][1] = min(r[i][1], r[i-1][1] + distance).
 * 4. Backward Pass: Pass backwards to propagate constraints from right neighbors back to left: 
 * r[i][1] = min(r[i][1], r[i+1][1] + distance).
 * 5. Peak Maximization: Between any two restricted buildings, the height can grow to a peak 
 * and drop down. The theoretical peak height between two points (id1, h1) and (id2, h2) is 
 * given by the math formula: floor((h1 + h2 + (id2 - id1)) / 2). Track the maximum peak found.
 */
import java.util.Arrays;
import java.util.Scanner;

public class MaxBuildingHeight1840 {
    public int maxBuilding(int n, int[][] restrictions) {
        int m = restrictions.length;
        // Create an extended array to include building 1 and building n as base boundaries
        int[][] r = new int[m + 2][2];
        
        for (int i = 0; i < m; i++) {
            r[i] = restrictions[i];
        }
        // Base constraint: building 1 must start at height 0
        r[m] = new int[]{1, 0};
        // End constraint: building n can at most reach height n - 1
        r[m + 1] = new int[]{n, n - 1};
        
        // Sort restrictions chronologically by building ID
        Arrays.sort(r, (a, b) -> Integer.compare(a[0], b[0]));
        
        // Pass 1: Propagate constraints from left to right
        for (int i = 1; i < r.length; i++) {
            r[i][1] = Math.min(r[i][1], r[i - 1][1] + (r[i][0] - r[i - 1][0]));
        }
        
        // Pass 2: Propagate constraints from right to left
        for (int i = r.length - 2; i >= 0; i--) {
            r[i][1] = Math.min(r[i][1], r[i + 1][1] + (r[i + 1][0] - r[i][0]));
        }
        
        int maxHeight = 0;
        // Calculate maximum peak heights between adjacent restricted zones
        for (int i = 0; i < r.length - 1; i++) {
            int id1 = r[i][0], h1 = r[i][1];
            int id2 = r[i + 1][0], h2 = r[i + 1][1];
            
            // Intersection peak math formula: peak = ((h1 + h2) + distance) / 2
            int peak = (int) (((long) h1 + h2 + (id2 - id1)) / 2);
            maxHeight = Math.max(maxHeight, peak);
        }
        
        return maxHeight;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter total number of buildings n:");
        int n = scanner.nextInt();
        
        System.out.println("Enter number of restrictions:");
        int m = scanner.nextInt();
        int[][] restrictions = new int[m][2];
        
        System.out.println("Enter restrictions (id height) line by line:");
        for (int i = 0; i < m; i++) {
            restrictions[i][0] = scanner.nextInt();
            restrictions[i][1] = scanner.nextInt();
        }
        
        MaxBuildingHeight1840 solver = new MaxBuildingHeight1840();
        int result = solver.maxBuilding(n, restrictions);
        System.out.println("Maximum possible building height: " + result);
        
        scanner.close();
    }
}